django/thunderstore/api/cyberstorm/views/packages.py

```diff
@@ -64,92 +61,110 @@ class PackageListResponseSerializer(serializers.Serializer):
+        attribute) and page_size (via page_size_query_param attribute).
+        The former would clash with the page field defined in our
+        PackageListRequestSerializer, breaking the API documentation
+        generated by drf_yasg. Returning an empty list means we have the
+        control and responsibility to make sure the documentation
+        matches reality.
+        """
+        return []
+
+
+class BasePackageListAPIView(ListAPIView):
     """
     Base class for community-scoped, paginated, filterable package listings.
 
     Classes implementing this base class should receive `community_id`
-    url parameter and implement the abstract methods listed below.
+    url parameter.
+
+    Methods with names starting prefixed with underscore are you custom
+    methods, whereas the rest are overwritten methods from ListAPIView.
     """
 
-    page_size = 20
+    pagination_class = PackageListPaginator
+    serializer_class = CyberstormPackagePreviewSerializer
+    viewname: str = ""  # Define in subclass
 
-    @abstractmethod
-    def _get_paginator_cache_key(self) -> str:
-        """
-        Return cache_key argument for CachedPaginator initialization.
-        """
-        raise NotImplementedError(
-            "PackageListApiView must implement _get_paginator_cache_key",
-        )
+    def get(self, request, *args, **kwargs):
+        return super().get(request, *args, **kwargs)
 
-    @abstractmethod
-    def _get_paginator_cache_vary_prefix(self) -> str:
-        """
-        Return endpoint and URL parameter specific string.
+    def list(self, request, *args, **kwargs):  # noqa: A003
+        assert self.paginator is not None
 
-        This is used as a part to create the cache_vary argument for
-        CachedPaginator initialization. Do not include query parameter
-        related information in this part.
-        """
-        raise NotImplementedError(
-            "PackageListApiView must implement _get_paginator_cache_vary_prefix",
-        )
+        queryset = self.filter_queryset(self.get_queryset())
+        page = self.paginate_queryset(queryset)
+        serializer = self.get_serializer(page, many=True)
+        response = self.paginator.get_paginated_response(serializer.data)
 
-    @abstractmethod
-    def _get_request_path(self) -> str:
-        """
-        Return path part of URL to access the endpoint
-        """
-        raise NotImplementedError(
-            "PackageListApiView must implement _get_request_path",
-        )
+        # Paginator's default implementation uses the Request object to
+        # construct previous/next links, which can open attack vectors
+        # via cache. Ideally this would have been overridden in the
+        # paginator itself, but that would require passing extra args,
+        # which would change the methods signatures, which is icky and
+        # not liked by MyPy either.
+        (previous_url, next_url) = self._get_sibling_pages()
```

Comment:
`_get_sibling_pages` seems like a function that should take in the current state it builds the next/prev URL based on as parameters to make the flow a bit more obvious. Reading this file from top to down currently on first read, at this point I can only assume it's accessing the relevant info via `self.request`, which if true could be made more obvious by passing it in as a parameter (most Django view functions pass around the request for example even if they could implicitly use `self.request` anyway).

Let's see if my assumption is correct once I get that far

Not a blocker

Reply:
Seems like my assumption was right, but yeah it'd be cool if I didn't have to make assumptions when reviewing

Reply:
I aim to make this DRFish, not Djangoish, and DRF accesses `self.request` quite frequently. And I could pass something typed "Request" into the functions but then you'd have to jump around to see what it is that's actually passed in since Request is such a common term. Or you could assume that it's DRF's Request object, but having to assume is not cool so I don't want to do that to you. (wontfix)